messages/block_hash_publisher/message_manager.go

```diff
@@ -0,0 +1,170 @@
+package block_hash_publisher
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp"
+	"github.com/ava-labs/awm-relayer/config"
+	"github.com/ava-labs/awm-relayer/vms"
+	"github.com/ava-labs/awm-relayer/vms/vmtypes"
+	teleporter_block_hash "github.com/ava-labs/teleporter/abi-bindings/Teleporter/TeleporterBlockHashReceiver"
+	"github.com/ethereum/go-ethereum/common"
+	"go.uber.org/zap"
+)
+
+const (
+	publishBlockHashGasLimit = 275000
+)
+
+type destinationSenderInfo struct {
+	client  vms.DestinationClient
+	address common.Address
+
+	useTimeInterval     bool
+	timeIntervalSeconds uint64
+	blockInterval       uint64
+
+	lastTimeSent uint64
+	lastBlock    uint64
+}
+
+func (d *destinationSenderInfo) shouldSend(blockTimestamp uint64, blockNumber uint64) bool {
+	if d.useTimeInterval {
+		interval := d.timeIntervalSeconds
+		if time.Unix(int64(blockTimestamp), 0).Sub(time.Unix(int64(d.lastTimeSent), 0)) < (time.Duration(interval) * time.Second) {
+			return false
+		}
+	} else {
+		interval := d.blockInterval
+		if blockNumber-d.lastBlock < interval {
+			return false
+		}
+	}
+	return true
+}
+
+type messageManager struct {
+	destinations map[ids.ID]*destinationSenderInfo
+	logger       logging.Logger
+}
+
+func NewMessageManager(
+	logger logging.Logger,
+	messageProtocolAddress common.Hash,
+	messageProtocolConfig config.MessageProtocolConfig,
+	destinationClients map[ids.ID]vms.DestinationClient,
+) (*messageManager, error) {
+	// Marshal the map and unmarshal into the Block Hash Publisher config
+	data, err := json.Marshal(messageProtocolConfig.Settings)
+	if err != nil {
+		logger.Error("Failed to marshal Block Hash Publisher config")
+		return nil, err
+	}
+	var messageConfig Config
+	if err := json.Unmarshal(data, &messageConfig); err != nil {
+		logger.Error("Failed to unmarshal Block Hash Publisher config")
+		return nil, err
+	}
+
+	if err := messageConfig.Validate(); err != nil {
+		logger.Error(
+			"Invalid Block Hash Publisher config.",
+			zap.Error(err),
+		)
+		return nil, err
+	}
+
+	destinations := make(map[ids.ID]*destinationSenderInfo)
+	for _, destination := range messageConfig.DestinationChains {
+		destinationID, err := ids.FromString(destination.ChainID)
+		if err != nil {
+			logger.Error(
+				"Failed to decode base-58 encoded destination chain ID",
+				zap.Error(err),
+			)
+			return nil, err
+		}
+		destinations[destinationID] = &destinationSenderInfo{
+			useTimeInterval:     destination.useTimeInterval,
+			timeIntervalSeconds: destination.timeIntervalSeconds,
+			blockInterval:       destination.blockInterval,
+			address:             common.HexToAddress(destination.Address),
```

Comment:
Can you confirm that destination.Address has already been verified so HexToAddress should never fail?